narwhals/functions.py

```diff
@@ -97,6 +98,19 @@ def from_dict(
         │ 2   ┆ 4   │
         └─────┴─────┘
     """
+    from narwhals.series import Series
+    from narwhals.translate import to_native
+
+    if not data:
+        msg = "from_dict cannot be called with empty dictionary"
+        raise ValueError(msg)
+    if native_namespace is None:
+        val = next(iter(data.values()))
+        if not isinstance(val, Series):
+            msg = "Calling `from_dict` without `native_namespace` is only supported if all input values are already Narwhals Series"
+            raise TypeError(msg)
+        native_namespace = val.__native_namespace__()
+        data = {key: to_native(value) for key, value in data.items()}
```

Comment:
Only addition I would consider is:

- either a test to break if values have different backends
- explicitly raise if that's the case